File: packages/core/src/driver.ts
import type { Batch, BatchResult } from "./batch";
import type { Capabilities } from "./capabilities";
import type { CollectionName, Key, Namespace } from "./key";
import type { DeleteOptions, ReadOptions, ScanOptions, WriteOptions } from "./options";
import type { StoredValue } from "./value";

/** Low-level storage contract implemented by yugendb drivers. */
export interface Driver {
  /** Stable driver name. */
  readonly name: string;
  /** Returns the features supported by this driver. */
  capabilities(): Capabilities;

  /** Opens any resources required by the driver. */
  initialise(): Promise<void>;
  /** Closes resources owned by the driver. */
  finalise(): Promise<void>;

  /** Reads a stored value by namespace, collection, and key. */
  get(
    namespace: Namespace,
    collection: CollectionName,
    key: Key,
    options: ReadOptions,
  ): Promise<StoredValue | null>;

  /** Writes a stored value by namespace, collection, and key. */
  set(
    namespace: Namespace,
    collection: CollectionName,
    key: Key,
    value: StoredValue,
    options: WriteOptions,
  ): Promise<void>;

  /** Deletes a stored value by namespace, collection, and key. */
  delete(
    namespace: Namespace,
    collection: CollectionName,
    key: Key,
    options: DeleteOptions,
  ): Promise<boolean>;

  /** Returns whether a non-expired value exists. */
  exists(namespace: Namespace, collection: CollectionName, key: Key): Promise<boolean>;

  /** Scans stored values whose keys start with a prefix. */
  scanPrefix(
    namespace: Namespace,
    collection: CollectionName,
    options: ScanOptions,
  ): Promise<Array<[Key, StoredValue]>>;

  /** Applies a batch of write operations. */
  batch(batch: Batch): Promise<BatchResult>;
}

File: crates/yugendb-core/src/driver.rs
//! Driver trait for concrete storage backends.

use async_trait::async_trait;

use crate::{
    Batch, BatchResult, Capabilities, CollectionName, DeleteOptions, Key, Namespace, ReadOptions,
    Result, ScanOptions, StoredValue, WriteOptions,
};

/// Async storage driver contract.
#[async_trait]
pub trait Driver: Send + Sync {
    /// Stable driver name.
    fn name(&self) -> &'static str;

    /// Reports supported driver behaviour.
    fn capabilities(&self) -> Capabilities;

    /// Prepares the driver for use.
    async fn initialise(&self) -> Result<()>;

    /// Releases driver resources.
    async fn finalise(&self) -> Result<()>;

    /// Reads a stored value.
    async fn get(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        options: &ReadOptions,
    ) -> Result<Option<StoredValue>>;

    /// Writes a stored value.
    async fn set(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        value: StoredValue,
        options: &WriteOptions,
    ) -> Result<()>;

    /// Deletes a stored value and returns whether a value was removed.
    async fn delete(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        options: &DeleteOptions,
    ) -> Result<bool>;

    /// Returns whether a stored value exists.
    async fn exists(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
    ) -> Result<bool>;

    /// Scans values by key prefix.
    async fn scan_prefix(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        options: &ScanOptions,
    ) -> Result<Vec<(Key, StoredValue)>>;

    /// Applies a batch of write operations.
    async fn batch(&self, batch: Batch) -> Result<BatchResult>;
}
